Change AccessibilityState to std::optional<bool>

dnhan1707 opened the pull request:
## Summary:
Possibly solve issue #46988 

This allow the Native side distinguish between boolean and undefine state.
The boolean value indicate the component is selected or not
The undefined indicate that the component is not selectable. However since the Native side (before) only aware of boolean
With this change Native side will regconize the undefine state (not just boolean)

## Changelog:

[GENERAL] [CHANGED] - Changed AccessibilityState to std::optional<bool> with default std::nullopt

## Test Plan:

![Screenshot 2024-10-29 190734](https://github.com/user-attachments/assets/11eb69f3-cc2a-4379-ac72-0da64c15e094)

tdn120 commented:
@dnhan1707 I'm looking to pull this in, but it seems a bit odd that this is the only change.  Can you confirm this is the full functionality you need?

dnhan1707 commented:
> @dnhan1707 I'm looking to pull this in, but it seems a bit odd that this is the only change. Can you confirm this is the full functionality you need?
Hey, let's me double check the code base again to is any place that need to be change.
I saw that the `expanded` attribute is also set to `std::optional` in that same Struct
![image](https://github.com/user-attachments/assets/47354857-7edf-450a-8c06-31839a5998b9)

Let me compare how the react native is handling expanded compare with other attribute.

dnhan1707 commented:
> @dnhan1707 I'm looking to pull this in, but it seems a bit odd that this is the only change. Can you confirm this is the full functionality you need?


Hey @tdn120 , I updated something, it probably still work the same way
I added more handle case to prevent set tag and set state for native side
 means no more setting true/false when the acccessibilityState is empty but not null. This way, the Native side still ensure the default type is std::nullopt when it is empty.

Hope it make sense

dnhan1707 commented:
Hi @tdn120,  I saw that the Facebook Internal Linter and Build, Tests are failed.
- For the Linter, I ran `yarn perttier` and `yarn format-check`, and everything passed on my side

It would be great if you can revew the code and provide me somefeed backs on why it failed or how to improve.

tdn120 commented:
Lint is unhappy about Java formatting.  That's not really an issue, a quick run of our google-java-format config cleans those up.

For the failures, what I'm seeing are the usages of `isSelected` are now broken due to `std::optional<bool>` being incompatible with `bool`.

There probably needs to be some more discussion in #46988 on how we want to implement this change so that it doesn't break existing usages, or get consensus that it's worth breaking existing usages.

dnhan1707 commented:
> Lint is unhappy about Java formatting. That's not really an issue, a quick run of our google-java-format config cleans those up.
> 
> For the failures, what I'm seeing are the usages of `isSelected` are now broken due to `std::optional<bool>` being incompatible with `bool`.
> 
> There probably needs to be some more discussion in #46988 on how we want to implement this change so that it doesn't break existing usages, or get consensus that it's worth breaking them.

Thank you for your response, I'll absolutely look into it more
